src/value.rs

```diff
@@ -169,7 +169,26 @@ impl EntryReader {
     }
 
     /// Entry returns header, key and value.
-    pub fn entry<R: Read + Seek>(&mut self, _reader: &mut BufReader<R>) -> Result<EntryRef> {
-        unimplemented!()
+    pub fn entry<R: Read + Seek>(&mut self, reader: &mut BufReader<R>) -> Result<EntryRef> {
+        // first, read MAX_HEADER_SIZE bytes to buf
+        let size = reader.read(&mut self.buf)?;
```

Comment:
What if the reader reads an part of the data?

Reply:
What about using `read_exact` here and ignore EOF here? The only possibility that we read less than `MAX_HEADER_SIZE` is when it comes to end of file. If so, we decode the entry normally. For other errors, just panic.

Reply:
Is it possible to know the length ahead of calling the function?

Reply:
Value log or WAL is simply a sequence of records, where each record begins with a variable-size header and variable-size key + value. So it's impossible to know how large is header before actually reading from that. Meanwhile, we could also use an alternative encoding different from badger (e.g. fixed-size header)

Reply:
I have an idea. If you implement `Buf` for a `BufReader` using a wrapper, you can reuse `prost::decode_length_delimiter` and `prost::encode_length_delimiter` to decode the header.

Reply:
Now we use `prost::encoding::decode_varint` for this operation (which takes a `Buf`). In next few PRs, I would phase out `binary.rs` and use utilities provided by prost.